### PR TITLE
mruby consistent fiber handling

### PR DESCRIPTION
**Pull request by i110**
Currently `http_request` method can be never called because it depends on generator and requests.
This PR makes us able to call http_request and other asynchronous functions from outside of the rack handler.

We can do like the following:
```ruby
mruby.handler: |
  body = http_request("http://example.com/").join[2].join
  Proc.new do |env|
    [200, {}, [body]]
  end
```

There are two FIXMEs:
- [x] find better way to evaluate ruby code (https://github.com/mruby/mruby/issues/3415)
- [x] address fastpath issue (will do in another PR)

### CODE REVIEW

File: lib/handler/mruby.c [hunk 1]
```diff
@@ -329,6 +312,24 @@ static h2o_mruby_shared_context_t *get_shared_context(h2o_context_t *ctx)
...
+    h2o_mruby_config_vars_t config = ctx->handler->config;
+    mrb_value conf = mrb_hash_new_capa(mrb, 3);
+    mrb_hash_set(mrb, conf, mrb_symbol_value(mrb_intern_lit(mrb, "code")), mrb_str_new(mrb, config.source.base, config.source.len));
+    mrb_hash_set(mrb, conf, mrb_symbol_value(mrb_intern_lit(mrb, "file")), mrb_str_new(mrb, config.path, strlen(config.path)));
+    mrb_hash_set(mrb, conf, mrb_symbol_value(mrb_intern_lit(mrb, "line")), mrb_fixnum_value(config.lineno));
+
+    /* run code and generate handler */
+    mrb_value result =
+        mrb_funcall(mrb, mrb_obj_value(mrb->kernel_module), "_h2o_prepare_app", 1, conf);
+    assert(mrb_array_p(result));
```
Comment (kazuho) [inline lib/handler/mruby.c]: I believe that we cannot have the assertion here.

My understanding that we need to check test if `mrb->exc` is NULL or not and take appropriate actions (which is done in the caller) before asserting that the value being returned is an array.

File: lib/handler/mruby.c [hunk 2]
```diff
@@ -720,85 +747,106 @@ void h2o_mruby_run_fiber(h2o_mruby_generator_t *generator, mrb_value receiver, m
...
+            goto GotException;
+        } else if (status == H2O_MRUBY_CALLBACK_ID_CONFIGURING_APP) {
+            mrb_value pending = mrb_ary_new_capa(mrb, 2);
+            mrb_ary_set(mrb, pending, 0, receiver);
+            mrb_ary_set(mrb, pending, 1, input);
+            mrb_ary_push(mrb, ctx->pendings, pending);
+            goto Exit;
+        } else if (status == H2O_MRUBY_CALLBACK_ID_CONFIGURED_APP) {
+            mrb_int len = mrb_ary_len(mrb, ctx->pendings);
+            for (i = 0; i != len; ++i) {
```
Comment (kazuho) [inline lib/handler/mruby.c]: Could we declare `i` along with `len` since (from what I understand) this for loop is the only location that we use the local variable?

File: lib/handler/mruby.c [hunk 3]
```diff
@@ -720,85 +747,106 @@ void h2o_mruby_run_fiber(h2o_mruby_generator_t *generator, mrb_value receiver, m
...
-                goto GotException;
+        if (status == H2O_MRUBY_CALLBACK_ID_EXCEPTION_RAISED) {
+            mrb->exc = mrb_obj_ptr(mrb_ary_entry(output, 1));
+            generator = h2o_mruby_get_generator(mrb, mrb_ary_entry(output, 2));
+            goto GotException;
+        } else if (status == H2O_MRUBY_CALLBACK_ID_CONFIGURING_APP) {
+            mrb_value pending = mrb_ary_new_capa(mrb, 2);
+            mrb_ary_set(mrb, pending, 0, receiver);
+            mrb_ary_set(mrb, pending, 1, input);
+            mrb_ary_push(mrb, ctx->pendings, pending);
```
Comment (kazuho) [inline lib/handler/mruby.c]: Could you please elaborate how `ctx->pendings` is protected from GC?

Comment (kazuho) [inline lib/handler/mruby.c]: PS. Or, if you would like, maybe we can have the pending list kept in the mruby-side. IIRC we have discussed the possibility before, but cannot recall why the idea was dismissed.

**Comment by yannick** [comment]
as discussed in slack i still see crashes.
config: `neverbleed: OFF` and a ecdsa-with-SHA256 prime256v1

```
/usr/bin/h2o(h2o_mruby_run_fiber+0xf0)[0x503ef0] h2o_mruby_run_fiber at ??:?
/usr/bin/h2o[0x5069d2] h2o_mruby_http_request_init_context at ??:?
h2o server (pid:5951) is ready to serve requests
/usr/bin/h2o[0x46b546] h2o_sem_set_capacity at ??:?
/usr/bin/h2o(redisProcessCallbacks+0xd9)[0x45bb59] redisProcessCallbacks at ??:?
/usr/bin/h2o[0x46d491] h2o_numproc at ??:?
/usr/bin/h2o[0x46d586] h2o_numproc at ??:?
/usr/bin/h2o(h2o_evloop_run+0x3d)[0x46fa7d] h2o_evloop_run at ??:?
/usr/bin/h2o[0x4e982f] neverbleed_init at ??:?
/usr/lib/libpthread.so.0(+0x72e7)[0x7f6c39ef72e7] ?? ??:0
/usr/lib/libc.so.6(clone+0x3f)[0x7f6c3917e54f] ?? ??:0
```

could that somehow be triggered by clients that can not cope with ecdsa ?

**Comment by kazuho** [comment]
@yannick Could it be the case that the error log contains stacktrace of multiple crashes? The trace contains call-paths that cannot happen (for example `h2o_numproc` never calls `redisProcessCallbacks`), and the most likely explanation is that the log is a mixture of multiple crashes.

If that is the case, I would suggest running the server in `--mode=worker` to prevent the worker process from automatically respawn, so that we can analyze a single crash.

@i110 Could it be the case that we are starting the execution of mruby code before configuration ends? I believe that we should delay the execution of mruby code using a zero-second timeout (relates to #1242).

**Comment by yannick** [comment]
@kazuho quite possible, i think it would be good to have an extra line and timestamps for crashes.

the worker is a bit of a problem, since i can't reproduce the error in synthetic tests and it needs a lot of live traffic to happen

**Comment by kazuho** [comment]
@yannick #1254 adds the process id to the stacktrace. I would appreciate it if you could give it a try.

**Comment by yannick** [comment]
@kazuho added all the latest commits and i do  not see any crashes anymore during the last day. looks good

File: lib/handler/mruby.c (hunk 3 above)
Comment (i110) [inline lib/handler/mruby.c]: > Could you please elaborate how ctx->pendings is protected from GC?

In `on_context_init` , `ctx->pendings` is allocated before `mrb_gc_arena_save` is called. So arena index always indicates the succeeding position after it, thus I believe `ctx->pendings` is protected from gc (if I understand mruby's gc correctly)

> PS. Or, if you like, maybe we can have the pending list kept in the mruby-side. IIRC we have discussed the possibility before, but cannot recall why the idea was dismissed.

IIRC the conclusion was "Either's fine", but then I came to feel that handling this in c is more simple and clear than doing in mruby (to handle it in mruby, we have to find a dirty way to pass the generator argument to app proc keeping rack interface, otherwise we need to check whether `app` is initialized or not in `runner` fiber, which may decrease runtime  performance, etc..)

**Comment by i110** [comment]
Sorry for my late response, I had been on vacation from the begging of April.

@kazuho I fixed 2 issues and replied to your comment.

> @i110 Could it be the case that we are starting the execution of mruby code before configuration ends? I believe that we should delay the execution of mruby code using a zero-second timeout (relates to #1242).

Could you elaborate why you thought that it's needed? In my quick looking, #1242 fixes the problems around communicating with other threads, but I believe that mruby handler is completely isolated from other threads.

@yannick you never see mruby crashes anymore, is it right? If so, the crashes was caused by other problems, not this PR?

**Comment by yannick** [comment]
@i110 yep, i don't see any new crashes after the last fixes.

File: lib/handler/mruby.c (hunk 1 above)
Comment (kazuho) [inline lib/handler/mruby.c]: No, my point was that the interface of `prepare_fibers` is designed to return without doing an assertion when the Rack handler fails to initialize. See how it is called in `on_context_init`:

```c
    mrb_value fibers = prepare_fibers(handler_ctx);
    if (mrb->exc != NULL) {
        fprintf(stderr, "mruby raised: %s\n", RSTRING_PTR(mrb_inspect(mrb, mrb_obj_value(mrb->exc))));
```

If we do an assert in `prepare_fiber`, then a user cannot see that the rack handler _the user wrote_ has failed to initialize.

File: lib/handler/mruby.c [hunk 4]
```diff
@@ -329,6 +312,25 @@ static h2o_mruby_shared_context_t *get_shared_context(h2o_context_t *ctx)
...
+    h2o_mruby_config_vars_t config = ctx->handler->config;
+    mrb_value conf = mrb_hash_new_capa(mrb, 3);
+    mrb_hash_set(mrb, conf, mrb_symbol_value(mrb_intern_lit(mrb, "code")), mrb_str_new(mrb, config.source.base, config.source.len));
+    mrb_hash_set(mrb, conf, mrb_symbol_value(mrb_intern_lit(mrb, "file")), mrb_str_new(mrb, config.path, strlen(config.path)));
+    mrb_hash_set(mrb, conf, mrb_symbol_value(mrb_intern_lit(mrb, "line")), mrb_fixnum_value(config.lineno));
+
+    /* run code and generate handler */
+    mrb_value result =
+        mrb_funcall(mrb, mrb_obj_value(mrb->kernel_module), "_h2o_prepare_app", 1, conf);
+    h2o_mruby_assert(mrb);
```
Comment (kazuho) [inline lib/handler/mruby.c]: Well, my point in the [previous comment](https://github.com/h2o/h2o/pull/1173#discussion_r110579168) was that the interface of `prepare_fibers` is designed to return without doing an assertion when the Rack handler fails to initialize. See how it is called in `on_context_init`:

```c
    mrb_value fibers = prepare_fibers(handler_ctx);
    if (mrb->exc != NULL) {
        fprintf(stderr, "mruby raised: %s\n", RSTRING_PTR(mrb_inspect(mrb, mrb_obj_value(mrb->exc))));
```

If we do an assert in `prepare_fiber`, then a user cannot see that the rack handler _the user wrote_ has failed to initialize.

File: lib/handler/mruby.c (hunk 3 above)
Comment (kazuho) [inline lib/handler/mruby.c]: Thank you for the clarification. Makes sense!

File: lib/handler/mruby.c (hunk 4 above)
Comment (i110) [inline lib/handler/mruby.c]: @kazuho Thank you for the clarification, I understand what you meant.
I rechecked `on_context_init`, then I came to believe that `_h2o_prepare_fibers` itself doesn't do nothing without creating fibers, so we need not to check exceptions and report to the users. So I actually remove exception check at https://github.com/h2o/h2o/pull/1173/commits/e592dcac51d4ca5d0b3376aa1288f51bd805801c. How's this?

**Comment by kazuho** [comment]
Thank you for working this huge change!

There are some leftovers (i.e. integrate with streaming request body) but it seems to make sense to merge now.
